File: 394-decode-string/decode-string.java
class Solution {
    public String decodeString(String s) {
        StringBuilder sb = new StringBuilder();
        Stack<StringBuilder> charSt = new Stack<>();
        Stack<Integer> countSt = new Stack<>();
        int curr = 0;

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if(Character.isDigit(ch)) {
                curr = curr * 10 + (ch - '0');
            } else if (ch == '[') {
                countSt.push(curr);
                curr = 0;
                charSt.push(sb);
                sb = new StringBuilder();
            } else if (ch == ']') {
                int k = countSt.pop();
                StringBuilder temp = sb;
                sb = charSt.pop();
                while (k > 0) {
                    k--;
                    sb.append(temp);
                }
            } else {
                sb.append(ch);
            }
        }
        return sb.toString();
    }
}

File: 394-decode-string/decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        res = []
        k = 0
        char_stack = []
        num_stack = []

        for char in s:
            if char.isdigit():
                k = k * 10 + int(char)
            elif char.isalpha():
                res.append(char)
            
            if char == "[":
                # store onto stack to pull later
                num_stack.append(k)
                char_stack.append("".join(res))

                # reset variables
                res = []
                k = 0

            elif char == "]":
                prefix = char_stack.pop()
                temp_num = num_stack.pop()

                inner_string = "".join(res)
                new_segment = prefix + (inner_string * temp_num)
                res = [new_segment]

        return "".join(res)